### PR TITLE
[Bug 1627769] Fix slow leak of wedged workers

### PR DESCRIPTION
**Pull request by imbstack**
Previous to this, the default lifecycles weren't getting applied. This was
fine in the 99% case where the worker succesfully registered as at that time
the `terminateAfter` date gets set to the `expires` of the creds returned.

However for workers that did not register, the `reregisterTimeout` was null
and these workers were allowed to live forever. To apply this to all currently
existing pools, we should run `ci-admin apply`/`tc-admin apply`.

We had to add a default to all places where the lifecycle `$ref` was used. However Ajv does not allow the default to be specified outside of `properties` and `items` so we just do it where it is "imported".

This also changes our reference validation because in Ajv, `$ref` doesn't actually have to be alone. It can have siblings that it will merge with.

Bugzilla Bug: [1627769](https://bugzilla.mozilla.org/show_bug.cgi?id=1627769)

### CODE REVIEW

**Comment by imbstack** [comment]
Yeah, the only thing about having defaults in two places is that i worried about making sure we would update both places later if we got around to it but that’s a good point I’ll add some defaults there too

**Comment by djmitche** [comment]
I *think* we've also tried to avoid defaults in schemas because their behavior is kind of wonky -- as you've seen here.  But I am just parroting jonas here, so who knows :)

**Comment by imbstack** [comment]
I think we fixed a lot of the wonkiness with the patch we made to it forever ago. The defaults are nice for docs at least. Would you rather I just remove the schema default entirely here and just have the default in the code or do both?

**Comment by djmitche** [comment]
Both seems good -- the schema defaults are just to `{}` whereas the code will apply the defaults for the particular properties, so I think they aren't overlapping too badly.  The critical change is to treat `undefined` as `{}` in the code.

**Comment by imbstack** [comment]
Interestingly setting the default `{}` allows the defaults of the lifecycle options themselves to take effect so the default is essentially `{reregistrationTimeout: NNNN}` but that makes sense. Will get something up shortly.

File: services/worker-manager/schemas/v1/worker-lifecycle.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ properties:
   reregistrationTimeout:
     title: Checkin Timeout
     type: integer
-    default: 345600   # 3 days
+    default: 345600   # 4 days (note this is also set in interpretLifecycle; update both if changing)
```
Comment (djmitche) [inline services/worker-manager/schemas/v1/worker-lifecycle.yml:24]: :+1:

**Comment by imbstack** [comment]
Oh wait, the default schema stuff is going to wreak havoc on tc-admin/ci-admin isn't it? I should just do the code-only bits I guess unless the admin tools understand this?

**Comment by djmitche** [comment]
I'm not sure how tc-admin or ci-admin would be affected -- neither tool uses the schemas, that I know of.  What are you seeing?

**Comment by imbstack** [comment]
I haven't tried it yet, I'm just thinking this won't work. Since the default is applied to the input, the the actual config that is stored changes and the tc-admin diff will always detect a change until the local config is updated as well.

**Comment by djmitche** [comment]
Oh, good thinking.  We could work around that in those tools, if it's worthwhile for the better display in the schema.

**Comment by imbstack** [comment]
Adding this to worker-lifecycle management sprint and we can decide what the nicest way to do that then. I'm leaning towards just having a default in the code at this point.

**Comment by imbstack** [comment]
ok, just gave up on making it show up in schemas. the schema docs will still show the default though

**Comment by djmitche** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1639365 filed for the intermittency

**Pull request merged**